### ISSUE TITLE
slower with GPU than with CPU

### ISSUE THREAD

**Issue by 9Oeufs** [alexjc/neural-doodle#82]
Hello,

First, thank you very much for this unbelievable tool. Congratulation.

I have a problem when using neural doodle with gpu. First, it is as slow as using gpu, even a bit slower (a few seconds on each 10 iterations). Second, it crashes at phase 2, out of memory.

I have a macbook pro retina with nvidia 750m and 2Go VRAM. I am pretty sure that my installation of cuda is good, but i don't know how to verify it's perfectly ok with theano. Is there any way to find if everything is ok for me?

thank you very much for your work

**Comment by alexjc** [comment]
Check with `nvidia-smi` what the GPU is doing while you're running. I'll try to update the documentation with more details, but there is another Issue that has a screenshot of the output of the tool that you should expect.

**Comment by 9Oeufs** [comment]
I'd like to, but there is no nvidia-smi on mac os x
i didn't find good solution to check what the gpu is doing for my os

**Comment by alexjc** [comment]
Oh, hmm. That's tough. I'm running on Mac OSX too but I don't have an NVIDIA graphics card.

**Comment by 9Oeufs** [comment]
I will try with this tool, but i think it won't be as effective as nvidia-smi
https://github.com/phvu/cuda-smi

**Comment by 9Oeufs** [comment]
Here is my terminal:
crash in phase 2 (/4) and according to cuda-smi, memory on gpu never exceed 20% and are most of the time around 10%. I don't know what the problem could be, because according to my mac, it's really the nvidia gpu that is in use, and not the intel chipset. And cuda seems to be well installed.

`pc15:~ MNv$ python3 /Users/MNv/Desktop/COURS/RNA/neural-doodle/doodle.py --style /Users/MNv/Desktop/COURS/RNA/neural-doodle/samples/lili_2.jpg --output /Users/MNv/Desktop/COURS/RNA/neural-doodle/samples/lili_3.png \
>                   --device=gpu0 --iterations=50 --phase=4 --save-every=10
Neural Doodle for semantic style transfer.
  - Using device `gpu0` for processing the images.
  - Adding `/Users/MNv/Desktop/COURS/RNA/neural-doodle/samples/lili_3_sem.png` as semantic map.
  - Loading `/Users/MNv/Desktop/COURS/RNA/neural-doodle/samples/lili_2.jpg` for style data.
  - Adding `/Users/MNv/Desktop/COURS/RNA/neural-doodle/samples/lili_2_sem.png` as semantic map.

Phase #0: resolution 64x82  scale 0.125
  - Style layer 3_1: (196, 260, 3, 3) patches in 458kb.
  - Style layer 4_1: (36, 516, 3, 3) patches in 167kb.

  0   loss 5.03e+03   style 3_1 8.16e+02  4_1 4.00e+03   quality  3.2%   time 1.4s 
 10   loss 2.27e+03   style 3_1 5.37e+02  4_1 1.60e+03   quality 33.8%   time 12.5s 
 20   loss 1.75e+03   style 3_1 4.79e+02  4_1 1.17e+03   quality 50.7%   time 13.0s 
 30   loss 1.47e+03   style 3_1 4.37e+02  4_1 9.57e+02   quality 60.0%   time 13.0s 
 40   loss 1.28e+03   style 3_1 3.92e+02  4_1 8.31e+02   quality 66.2%   time 13.0s 

Phase #1: resolution 128x164  scale 0.25
  - Style layer 3_1: (900, 260, 3, 3) patches in 2,106kb.
  - Style layer 4_1: (196, 516, 3, 3) patches in 910kb.

 50   loss 2.85e+03   style 3_1 5.08e+02  4_1 2.33e+03   quality  4.5%   time 5.3s 
 60   loss 1.24e+03   style 3_1 3.13e+02  4_1 9.22e+02   quality 39.9%   time 47.7s 
 70   loss 1.04e+03   style 3_1 2.67e+02  4_1 7.67e+02   quality 61.6%   time 47.2s 
 80   loss 9.43e+02   style 3_1 2.42e+02  4_1 6.97e+02   quality 74.6%   time 47.4s 
 90   loss 8.77e+02   style 3_1 2.22e+02  4_1 6.52e+02   quality 81.6%   time 47.0s 

Phase #2: resolution 256x328  scale 0.5
  - Style layer 3_1: (3906, 260, 3, 3) patches in 9,140kb.
  - Style layer 4_1: (900, 516, 3, 3) patches in 4,179kb.

100   loss 1.84e+03   style 3_1 2.68e+02  4_1 1.57e+03   quality  4.9%   time 19.2s 
Error allocating 193462272 bytes of device memory (out of memory). Driver report 170016768 bytes free and 2147024896 bytes total 
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/theano/compile/function_module.py", line 859, in __call__
    outputs = self.fn()
RuntimeError: GpuCorrMM failed to allocate working memory of 576 x 83968


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/MNv/Desktop/COURS/RNA/neural-doodle/doodle.py", line 651, in <module>
    generator.run()
  File "/Users/MNv/Desktop/COURS/RNA/neural-doodle/doodle.py", line 628, in run
    iprint=-1)                       # Handle our own logging of information.
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/scipy/optimize/lbfgsb.py", line 193, in fmin_l_bfgs_b
    **opts)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/scipy/optimize/lbfgsb.py", line 330, in _minimize_lbfgsb
    f, g = func_and_grad(x)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/scipy/optimize/lbfgsb.py", line 278, in func_and_grad
    f = fun(x, *args)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/scipy/optimize/optimize.py", line 289, in function_wrapper
    return function(*(wrapper_args + args))
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/scipy/optimize/optimize.py", line 63, in __call__
    fg = self.fun(x, *args)
  File "/Users/MNv/Desktop/COURS/RNA/neural-doodle/doodle.py", line 534, in evaluate
    grads, *losses = self.compute_grad_and_losses(current_img, self.content_map, *current_best)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/theano/compile/function_module.py", line 871, in __call__
    storage_map=getattr(self.fn, 'storage_map', None))
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/theano/gof/link.py", line 314, in raise_with_op
    reraise(exc_type, exc_value, exc_trace)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/theano/compile/function_module.py", line 859, in __call__
    outputs = self.fn()
RuntimeError: GpuCorrMM failed to allocate working memory of 576 x 83968

Apply node that caused the error: GpuCorrMM_gradInputs{(1, 1), (1, 1)}(GpuContiguous.0, GpuContiguous.0)
Toposort index: 229
Inputs types: [CudaNdarrayType(float32, 4D), CudaNdarrayType(float32, 4D)]
Inputs shapes: [(64, 64, 3, 3), (1, 64, 328, 256)]
Inputs strides: [(576, 9, 3, 1), (0, 83968, 256, 1)]
Inputs values: ['not shown', 'not shown']
Outputs clients: [[GpuElemwise{Composite{((i0 * i1) + (i0 * i1 * sgn(i2)))}}[(0, 1)](CudaNdarrayConstant{[[[[ 0.5]]]]}, GpuCorrMM_gradInputs{(1, 1), (1, 1)}.0, GpuElemwise{Add}[(0, 0)].0)]]

HINT: Re-running with most Theano optimization disabled could give you a back-trace of when this node was created. This can be done with by setting the Theano flag 'optimizer=fast_compile'. If that does not work, Theano optimizations can be disabled with 'optimizer=None'.
HINT: Use the Theano flag 'exception_verbosity=high' for a debugprint and storage map footprint of this apply node.
pc15:~ MNv$ 
`

**Comment by alexjc** [comment]
You're just out of memory for the next phase, memory increases a lot between phases. The next update will include better memory usage...

**Comment by 9Oeufs** [comment]
Ok, I understand that part but it doesn't explain why gpu is as slow as cpu, even with cuda and cudnn

**Comment by jwatte** [comment]
It could be that Apple drivers are slow and the mobile 750 is not a particularly powerful GPU. (In fact, Apple has no MacBook with a powerful GPU!)

Or it could be something different. What does a CPU profile say the system is doing while using the GPU? Spending lots of time waiting on the GPU or burning cycles in the driver?

**Comment by alexjc** [comment]
I've seen lots of swapping on bigger images, maybe that's also a problem. Do you find small images are significantly faster?

**Comment by 9Oeufs** [comment]
Ok I ran some more tests, I start to think that the macbook pro GPU is a shit.
First I discover that between 20% to 25% of my gpu memory is used ... when the gpu is not used (when mac os run on the intel iris) so I have only 1600 mo VRAM available

second I run a comparing test with the same pictures, with gpu and with cpu:

>
Phase #0: resolution 67x67  scale 0.25
  - Style layer 3_1: (225, 260, 3, 3) patches in 526kb.
  - Style layer 4_1: (36, 516, 3, 3) patches in 167kb.
>
  0   loss 7.44e+03   style 3_1 1.13e+03  4_1 6.10e+03   quality  1.6%   time 1.2s 
 10   loss 1.99e+03   style 3_1 6.50e+02  4_1 1.21e+03   quality 26.7%   time 9.9s 
 20   loss 1.35e+03   style 3_1 5.07e+02  4_1 7.49e+02   quality 38.7%   time 9.8s 
 30   loss 1.06e+03   style 3_1 4.32e+02  4_1 5.61e+02   quality 45.0%   time 9.8s 
>
Phase #1: resolution 135x135  scale 0.5
  - Style layer 3_1: (961, 260, 3, 3) patches in 2,248kb.
  - Style layer 4_1: (196, 516, 3, 3) patches in 910kb.
>
 40   loss 3.58e+03   style 3_1 9.54e+02  4_1 2.61e+03   quality  3.7%   time 3.7s 
 50   loss 1.31e+03   style 3_1 4.63e+02  4_1 8.36e+02   quality 34.7%   time 37.7s 
 60   loss 1.01e+03   style 3_1 3.85e+02  4_1 6.14e+02   quality 54.0%   time 38.6s 
 70   loss 8.54e+02   style 3_1 3.40e+02  4_1 5.06e+02   quality 65.1%   time 39.6s 
>
Phase #2: resolution 270x270  scale 1.0
  - Style layer 3_1: (4225, 260, 3, 3) patches in 9,886kb.
  - Style layer 4_1: (961, 516, 3, 3) patches in 4,462kb.
>
 80   loss 2.73e+03   style 3_1 6.67e+02  4_1 2.06e+03   quality  4.8%   time 15.9s 
 90   loss 1.22e+03   style 3_1 3.61e+02  4_1 8.59e+02   quality 41.2%   time 150.0s 
100   loss 9.35e+02   style 3_1 2.84e+02  4_1 6.50e+02   quality 62.7%   time 156.5s 
110   loss 8.01e+02   style 3_1 2.48e+02  4_1 5.52e+02   quality 74.7%   time 152.2s 
>
Optimization finished in 835.3s, average pixel error 9.8!
>

same on CPU:

>
Phase #0: resolution 67x67  scale 0.25
  - Style layer 3_1: (225, 260, 3, 3) patches in 526kb.
  - Style layer 4_1: (36, 516, 3, 3) patches in 167kb.
>
  0   loss 7.53e+03   style 3_1 1.13e+03  4_1 6.19e+03   quality  3.0%   time 1.1s 
 10   loss 2.02e+03   style 3_1 6.39e+02  4_1 1.25e+03   quality 27.2%   time 9.7s 
 20   loss 1.24e+03   style 3_1 4.58e+02  4_1 6.86e+02   quality 40.5%   time 9.1s 
 30   loss 8.94e+02   style 3_1 3.51e+02  4_1 4.76e+02   quality 48.4%   time 10.1s 
>
Phase #1: resolution 135x135  scale 0.5
  - Style layer 3_1: (961, 260, 3, 3) patches in 2,248kb.
  - Style layer 4_1: (196, 516, 3, 3) patches in 910kb.
>
 40   loss 3.18e+03   style 3_1 8.29e+02  4_1 2.34e+03   quality  3.9%   time 3.9s 
 50   loss 1.07e+03   style 3_1 3.70e+02  4_1 6.93e+02   quality 36.8%   time 36.4s 
 60   loss 8.12e+02   style 3_1 3.02e+02  4_1 5.04e+02   quality 55.7%   time 40.6s 
 70   loss 6.77e+02   style 3_1 2.59e+02  4_1 4.11e+02   quality 66.4%   time 39.5s 
>
Phase #2: resolution 270x270  scale 1.0
  - Style layer 3_1: (4225, 260, 3, 3) patches in 9,886kb.
  - Style layer 4_1: (961, 516, 3, 3) patches in 4,462kb.
>
 80   loss 2.53e+03   style 3_1 6.14e+02  4_1 1.91e+03   quality  4.8%   time 16.5s 
 90   loss 1.05e+03   style 3_1 3.12e+02  4_1 7.38e+02   quality 41.4%   time 154.0s 
100   loss 7.83e+02   style 3_1 2.37e+02  4_1 5.45e+02   quality 63.1%   time 144.7s 
110   loss 6.67e+02   style 3_1 2.04e+02  4_1 4.61e+02   quality 74.7%   time 142.3s 
>
Optimization finished in 809.6s, average pixel error 9.3!
>

So it run faster and is more efficient on cpu than on gpu. I don't understand, even if macbook's gpu is shitty, how can it be so useless?

And to check what cpu is doing during this operation, when running on cpu it is at more or less 120% cpu usage (I don't understand what it means btw, but it's data from activity monitor) and almost 600% max sometimes. Real memory size : 700mo - Virtual memory size 3Go

and when running on gpu, CPU usage is 100% - real memory size 700mo too - virtual memory size 30GO (!!!!)

If it say something to someone, I am not enough power user to understand if there is a problem somewhere or if this is normal

**Comment by alexjc** [comment]
When I test my code I test on the CPU on MacBook Pro, and I'm surprised how well the CPU does. I don't have a GPU though, but there's some kind of parallelization going on with CPU libraries under the hood of `Theano` I think.

**Comment by jwatte** [comment]
"when running on gpu, CPU usage is 100%"

That likely means that the Apple GPU drivers are inefficient, and the CPU can't drive the GPU fast enough for it to work out. Thus, you're not so much limited on the GPU, as you are on the ability of the Apple GPU drivers to efficiently push the work to the GPU and read it back again.

**Comment by 9Oeufs** [comment]
Yep, I think we reach the limits of the poor 750m and apple drivers.
But when I use neural style transfer https://github.com/jcjohnson/neural-style working on gpu is sinificantly faster. But I am stuck with the 2Go of VRAM of the GPU.

Macbook are really good laptop, but (like almost all laptops) not made for that kind of usage.
